Add http example

thisthat opened the pull request.

objectiser committed:
Set the type of the tags within the Jaeger exporter (#669)

* Set the type of the tags within the Jaeger exporter

Signed-off-by: Gary Brown <gary@brownuk.com>

* Address comment

Signed-off-by: Gary Brown <gary@brownuk.com>

* Address comments

Signed-off-by: Gary Brown <gary@brownuk.com>

ptravers committed:
Add named meter and meter factory (#668)

* Add named meters with factory provider

* Fix documentation to instantiate a meter from a meter factory

* Fix documentation incorrectly referencing meter not meter factory

* Fix MeterFactor not Meter

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Fix MeterFactor not Meter

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Fix MeterFactory not Meter

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Fix over 100 characters in line

* Refactor meter example names to my_library_example

bogdandrutu committed:
Rename MeasureBatchRecorder to BatchRecorder to allow other types to be recorded. (#675)

* Rename MeasureBatchRecorder to BatchRecorder to allow other types to be recorded.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* ./gradlew goJF

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Start 0.3.0 development cycle (#676)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Move InstrumentationLibraryInfo to common to be shared with metrics. (#682)

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

bogdandrutu committed:
Add LabelSet to the Meter API and change GetHandle to use it. (#681)

* Add LabelSet to the Meter API and change GetHandle to use it.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Add clarification in LabelSet javadoc

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

* Update api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Update api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Update api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

* Update api/src/main/java/io/opentelemetry/metrics/DefaultMeter.java

Co-Authored-By: Armin Ruech <armin.ruech@gmail.com>

carlosalberto committed:
Allow recording attributes in Span.Builder (#685)

* Initial support for attributes set in Span.Builder.

carlosalberto committed:
Handle null parent's SpanContextShim in the OT Shim. (#686)

SpanContextShim objects are created on demand,
as it's an expensive operation because of a global
write lock, hence we need to handle the case
of starting a Span whose parent has no SpanContextShim
yet.

Oberon00 committed:
Javadoc: Tracer is not simple (#687)

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpClient.java:
```diff
@@ -0,0 +1,116 @@
+  // Export traces in memory
+  InMemorySpanExporter inMemexporter = InMemorySpanExporter.create();
+  // Share context via http headers
+  HttpTextFormat<SpanContext> textFormat = new HttpTraceContext();
```
This should be a concern of the tracer so I'd defer the initialization to `initTracer()` which would assign `tracer.getTextFormat()` or not a keep a reference at all and just use `tracer.getTextFormat()` on demand.

arminru replied:
ping

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpClient.java:
```diff
@@ -0,0 +1,116 @@
+    // Start a span
+    Span span = tracer.spanBuilder("Request hello API").startSpan();
+    // Inject the request with the context
+    textFormat.inject(span.getContext(), con, setter);
```
```suggestion
    tracer.getTextFormat().inject(span.getContext(), con, setter);
```

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpClient.java:
```diff
@@ -0,0 +1,116 @@
+
+    // Process the request
+    con.setRequestMethod("GET");
+    int status = con.getResponseCode();
```
This should be embedded in a `try` block and a `catch` block should be used for reporting errors by setting the span status accordingly.

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpServer.java:
```diff
@@ -0,0 +1,165 @@
+
+public class HttpServer {
+
+  private class helloHandler implements HttpHandler {
```
```suggestion
  private class HelloHandler implements HttpHandler {
```

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpServer.java:
```diff
@@ -0,0 +1,165 @@
+        System.out.println(e.getMessage());
+      } finally {
+        // Close the span
+        span.end();
```
Don't forget to set the span status on success/catch.

arminru replied:
ping

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpServer.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    private SpanContext extractCtx(HttpExchange he) {
+      SpanContext ctx = (SpanContext) textFormat.extract(he, getter);
```
I'd do that right in `handle()` for this example since it's pretty concise anyway.

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpServer.java:
```diff
@@ -0,0 +1,165 @@
+          "client.info",
+          AttributeValue.stringAttributeValue(he.getRemoteAddress().getHostString()));
+      // Attach the event to the Span
+      span.addEvent("event", event);
```
The client info would be a span attribute rather than an event.
A better example would be something like "start processing" (which would make even more sense if that happened a bit after starting the span but it I think it's good enough for this example).

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpServer.java:
```diff
@@ -0,0 +1,165 @@
+    initTracer();
+    server = com.sun.net.httpserver.HttpServer.create(new InetSocketAddress(port), 0);
+    // Test urls
+    server.createContext("/", new helloHandler());
```
```suggestion
    server.createContext("/", new HelloHandler());
```

arminru reviewed (changes requested):
Please add span attributes according to the semantic conventions specified here:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-span-general.md
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpClient.java:
```diff
@@ -0,0 +1,136 @@
+    // See: https://github.com/open-telemetry/opentelemetry-specification/issues/270
+    Span span =
+        tracer.spanBuilder("/").setSpanKind(Span.Kind.CLIENT).startSpan();
+    // // TODO provide semantic convention attributes to Span.Builder
```
```suggestion
    // TODO provide semantic convention attributes to Span.Builder
```

arminru commented on examples/http/src/main/java/io/opentelemetry/example/HttpClient.java:
```diff
@@ -0,0 +1,136 @@
+       - http.scheme, peer.hostname, peer.port, http.target
+       - http.scheme, peer.ip, peer.port, http.target
+    */
+    span.setAttribute("http.url", "http://127.0.0.1:" + port);
```
you could also just use `url.toString`

z1c0 reviewed (approved):
LGTM!

arminru reviewed (approved):
👌 :shipit:

sano307 committed:
Fix the warning that occurs when a code is used in a code tag (#689)

Add build scripts for HTTP example

Add HTTP example

Improved naming and formatting

Add Semantic Convention to traces

Fix semantic convention

sano307 committed:
Fix the warning that occurs when a code is used in a code tag (#689)

thisthat committed:
Add Examples as OpenTelemetry project

thisthat committed:
Add Examples as OpenTelementry project

thisthat committed:
Removed Error Prone warnings

thisthat committed:
Removed Error Prone warnings

thisthat committed:
Fix Java Format

thisthat committed:
Fix Java Format

thisthat committed:
Fix Checkstyle errors

thisthat committed:
Fix Checkstyle errors

thisthat committed:
Merge branch 'add-http-example' of https://github.com/dynatrace-oss-contrib/opentelemetry-java into add-http-example

thisthat committed:
Changed the build configuration - Examples are compiled with the main project

thisthat committed:
Addressing comment of @carlosalberto

thisthat committed:
Addressed the feedback of @bogdandrutu

thisthat committed:
Fix Format
